In [2]:
import os
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
import tensorflow as tf
import random
%matplotlib inline

pd.set_option('display.max_rows', 200)
pd.set_option('display.max_columns', 200)

2023-11-01 15:34:27.802552: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [15]:

#Read dataset
train_a = pd.read_parquet('A/train_targets.parquet')
train_b = pd.read_parquet('B/train_targets.parquet')
train_c = pd.read_parquet('C/train_targets.parquet')
X_train_estimated_a = pd.read_parquet('A/X_train_estimated.parquet')
X_train_estimated_b = pd.read_parquet('B/X_train_estimated.parquet')
X_train_estimated_c = pd.read_parquet('C/X_train_estimated.parquet')
X_train_observed_a = pd.read_parquet('A/X_train_observed.parquet')
X_train_observed_b = pd.read_parquet('B/X_train_observed.parquet')
X_train_observed_c = pd.read_parquet('C/X_train_observed.parquet')

#add location to each sample
train_a["location"] = "A"
train_b["location"] = "B"
train_c["location"] = "C"
X_train_estimated_a["location"] = "A"
X_train_estimated_b["location"] = "B"
X_train_estimated_c["location"] = "C"
X_train_observed_a["location"] = "A"
X_train_observed_b["location"] = "B"
X_train_observed_c["location"] = "C"

#remove extra minute 00 sample
X_train_observed_a = X_train_observed_a.iloc[:-1,:]
X_train_observed_b = X_train_observed_b.iloc[:-1,:]
X_train_observed_c = X_train_observed_c.iloc[:-1,:]

#add date_calc column same as date_forecast column to observed data
X_train_observed_a.insert(0, "date_calc", X_train_observed_a["date_forecast"])
X_train_observed_b.insert(0, "date_calc", X_train_observed_b["date_forecast"])
X_train_observed_c.insert(0, "date_calc", X_train_observed_c["date_forecast"])

#concat all the samples and remove date_calc column
X_train_raw = pd.concat([X_train_observed_a,
                     X_train_observed_b,
                     X_train_observed_c,
                     X_train_estimated_a,
                     X_train_estimated_b,
                     X_train_estimated_c])

#feature indicating time between date_calc and date_forecast
X_train_raw["calc_time"] =(X_train_raw["date_forecast"] - X_train_raw["date_calc"]).astype('timedelta64[s]')

#fill nans
X_train_raw["snow_density:kgm3"] = X_train_raw["snow_density:kgm3"].apply(
    lambda a : np.isnan(a)
    ).map({True: 0, False: 1})

X_train_raw["ceiling_height_agl:m"] = X_train_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)

X_train_raw["cloud_base_agl:m"] = X_train_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)

#X_train_raw = X_train_raw.drop("msl_pressure:hPa", axis=1)
#X_train_raw = X_train_raw.drop("sfc_pressure:hPa", axis=1)
#X_train_raw = X_train_raw.drop("direct_rad_1h:J", axis=1)



#create seperate dataframes for measurments at minute 00, 15, 30 and 45
X_train00 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 0)].reset_index().iloc[:,1:]
X_train15 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 15)].reset_index().iloc[:,1:]
X_train30 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 30)].reset_index().iloc[:,1:]
X_train45 = X_train_raw[X_train_raw["date_forecast"].apply(lambda time: time.minute == 45)].reset_index().iloc[:,1:]

#remove redundant data
X_train15 = X_train15.iloc[:,2:-2]
X_train30 = X_train30.iloc[:,2:-2]
X_train45 = X_train45.iloc[:,2:-2]

#join observations into single sample
X_train = X_train00.join(X_train15, lsuffix="_00", rsuffix="_15").join(X_train30.join(X_train45, lsuffix="_30", rsuffix="_45"))

#rename column for merging with targets
X_train = X_train.rename(columns={"date_forecast" : "time"})

#concat target values and drop NaN values
targets = pd.concat([train_a,
                     train_b,
                     train_c]).dropna()

#merge weatherfeatures with corresponding target pv measurement
dataset = pd.merge(X_train, targets, how="right", on=["time", "location"])

#shuffle dataset
#dataset = dataset.sample(frac=1, random_state=43).reset_index().iloc[:,1:]

#sort by time
dataset = dataset.sort_values(by="time")

#split into features and targets
datasetX = dataset.iloc[:, :-1]
datasetY = dataset.iloc[:, -1:]

#add day_of_year and hour feature columns
datasetX["day"] = datasetX["time"].dt.day_of_year
datasetX["hour"] = datasetX["time"].dt.hour

#get indexes of samples in the months of the test dataset
indexMayJuneJuly = datasetX["time"].apply(lambda time : time.month in [5, 6, 7])

#OHE encoding for catagorical feature "location"
datasetX["location_A"] = datasetX["location"].apply(lambda a : a == "A").map({True: 1, False: 0})
datasetX["location_B"] = datasetX["location"].apply(lambda a : a == "B").map({True: 1, False: 0})
datasetX["location_C"] = datasetX["location"].apply(lambda a : a == "C").map({True: 1, False: 0})

datasetX = datasetX.drop("location", axis=1)

#move datecalc column
#date_calc_column = datasetX.pop("date_calc")
#datasetX.insert(183, "is_not_calculated", date_calc_column) #!constant 183 can be source of bugs

#fix nans for some reason???
#datasetX["is_not_calculated"] = datasetX["is_not_calculated"].fillna(method="ffill")

#map location labels to numbers
#datasetX["location"] = datasetX["location"].map({"A": 0, "B": 1, "C": 2})

#drop time and date_calc columns
datasetX = datasetX.iloc[:,2:]

#calculate mean and std for normalizing data, values should also be used for normalizing test data
dataMean = datasetX.mean()
dataStd = datasetX.std()

#normalize data
datasetX.iloc[:,:-1] = ((datasetX.iloc[:,:-1]-dataMean[:-1])/dataStd[:-1]).fillna(value=0)


evalsetSize = 10000
evalsetIndex = 0
#partition into training and evalset
trainsetX = pd.concat([datasetX.iloc[:evalsetIndex,:], datasetX.iloc[evalsetIndex+evalsetSize:,:]])
trainsetY = pd.concat([datasetY.iloc[:evalsetIndex,:], datasetY.iloc[evalsetIndex+evalsetSize:,:]])
#trainsetIndexMayJuneJuly = indexMayJuneJuly[:85000]

evalsetX = datasetX.iloc[evalsetIndex:evalsetIndex+evalsetSize,:]
evalsetY = datasetY.iloc[evalsetIndex:evalsetIndex+evalsetSize,:]
#evalsetIndexMayJuneJuly = indexMayJuneJuly[85000:]
dataset = dataset.reset_index().iloc[:,1:]
display(dataset[dataset["time"].apply(lambda a : a.month == 5 and a.day==1 and a.hour == 0)])
display(dataset.iloc[5200:])

/var/folders/bm/37km6rb530l8y0vv0znc0zb00000gn/T/ipykernel_54056/3903213449.py:130: PerformanceWarning: Adding/subtracting object-dtype array to TimedeltaArray not vectorized.
  datasetX.iloc[:,:-1] = ((datasetX.iloc[:,:-1]-dataMean[:-1])/dataStd[:-1]).fillna(value=0)


date_calc       time  absolute_humidity_2m:gm3_00  \
2880  2019-05-01 2019-05-01                          6.9   
25404 2020-05-01 2020-05-01                          3.5   
25405 2020-05-01 2020-05-01                          3.7   
25406 2020-05-01 2020-05-01                          3.7   
51677 2021-05-01 2021-05-01                          4.7   
51678 2021-05-01 2021-05-01                          4.8   
51679 2021-05-01 2021-05-01                          4.8   
77828 2022-05-01 2022-05-01                          6.1   
77829 2022-05-01 2022-05-01                          6.1   

       air_density_2m:kgm3_00  ceiling_height_agl:m_00  \
2880                    1.260               717.500000   
25404                   1.260             -1000.000000   
25405                   1.257             -1000.000000   
25406                   1.262             -1000.000000   
51677                   1.287             -1000.000000   
51678                   1.288             -1000.000000   
51679                   1.284             -1000.000000   
77828                   1.277              2282.199951   
77829                   1.279              2283.500000   

       clear_sky_energy_1h:J_00  clear_sky_rad:W_00  cloud_base_agl:m_00  \
2880                        0.0                 0.0           717.500000   
25404                       0.0                 0.0         -1000.000000   
25405                       0.0                 0.0         -1000.000000   
25406                       0.0                 0.0         -1000.000000   
51677                       0.0                 0.0         -1000.000000   
51678                       0.0                 0.0         -1000.000000   
51679                       0.0                 0.0         -1000.000000   
77828                       0.0                 0.0          2282.199951   
77829                       0.0                 0.0          2283.500000   

       dew_or_rime:idx_00  dew_point_2m:K_00  diffuse_rad:W_00  \
2880                  0.0         278.399994               0.0   
25404                 0.0         268.700012               0.0   
25405                 0.0         269.799988               0.0   
25406                 0.0         269.799988               0.0   
51677                -1.0         272.600006               0.0   
51678                 0.0         273.000000               0.0   
51679                 0.0         273.000000               0.0   
77828                 0.0         276.600006               0.0   
77829                 0.0         276.700012               0.0   

       diffuse_rad_1h:J_00  direct_rad:W_00  direct_rad_1h:J_00  \
2880                   0.0              0.0                 0.0   
25404                  0.0              0.0                 0.0   
25405                  0.0              0.0                 0.0   
25406                  0.0              0.0                 0.0   
51677                  0.0              0.0                 0.0   
51678                  0.0              0.0                 0.0   
51679                  0.0              0.0                 0.0   
77828                  0.0              0.0                 0.0   
77829                  0.0              0.0                 0.0   

       effective_cloud_cover:p_00  elevation:m_00  fresh_snow_12h:cm_00  \
2880                   100.000000             7.0                   0.0   
25404                   22.799999            24.0                   0.0   
25405                   25.400000             7.0                   0.0   
25406                   25.400000             6.0                   0.0   
51677                   76.000000            24.0                   0.0   
51678                   75.000000             6.0                   0.0   
51679                   75.000000             7.0                   0.0   
77828                   96.500000             7.0                   0.0   
77829                   96.500000             6.0        

date_calc                time  absolute_humidity_2m:gm3_00  \
5200  2019-07-04 19:00:00 2019-07-04 19:00:00                          6.9   
5201  2019-07-04 19:00:00 2019-07-04 19:00:00                          6.9   
5202  2019-07-04 20:00:00 2019-07-04 20:00:00                          7.2   
5203  2019-07-04 20:00:00 2019-07-04 20:00:00                          7.2   
5204  2019-07-04 21:00:00 2019-07-04 21:00:00                          7.2   
...                   ...                 ...                          ...   
93019 2023-04-29 07:00:05 2023-04-30 22:00:00                          4.5   
93020 2023-04-29 07:00:05 2023-04-30 22:00:00                          4.4   
93021 2023-04-29 07:00:05 2023-04-30 23:00:00                          4.5   
93022 2023-04-29 07:00:05 2023-04-30 23:00:00                          4.5   
93023 2023-04-29 07:00:05 2023-04-30 23:00:00                          4.4   

       air_density_2m:kgm3_00  ceiling_height_agl:m_00  \
5200                    1.232              1062.199951   
5201                    1.234              1063.000000   
5202                    1.237               950.700012   
5203                    1.238               951.700012   
5204                    1.241               840.200012   
...                       ...                      ...   
93019                   1.281              1532.900024   
93020                   1.279              1446.599976   
93021                   1.281              1054.199951   
93022                   1.281              1055.800049   
93023                   1.280               897.200012   

       clear_sky_energy_1h:J_00  clear_sky_rad:W_00  cloud_base_agl:m_00  \
5200              612038.375000          116.400002          1062.199951   
5201              612032.687500          116.400002          1063.000000   
5202              267224.406250           39.500000           950.700012   
5203              267225.312500           39.500000           951.700012   
5204               70764.203125            4.600000           840.200012   
...                         ...                 ...                  ...   
93019                  0.000000            0.000000          1532.900024   
93020                  0.000000            0.000000          1446.599976   
93021                  0.000000            0.000000          1054.199951   
93022                  0.000000            0.000000          1055.800049   
93023                  0.000000            0.000000           897.200012   

       dew_or_rime:idx_00  dew_point_2m:K_00  diffuse_rad:W_00  \
5200                  0.0         278.700012         42.799999   
5201                  0.0         278.700012         42.799999   
5202                  0.0         279.200012         20.000000   
5203                  0.0         279.200012         20.000000   
5204                  0.0         279.100006          4.100000   
...                   ...                ...               ...   
93019                 0.0         272.299988          0.000000   
93020                 0.0         271.899994          0.000000   
93021                 0.0         272.200012          0.000000   
93022                 0.0         272.200012          0.000000   
93023                 0.0         272.000000          0.000000   

       diffuse_rad_1h:J_00  direct_rad:W_00  direct_rad_1h:J_00  \
5200         220248.093750        36.500000       184228.203125   
5201         220308.093750        36.599998       184335.796875   
5202         113116.703125        12.700000        88647.101562   
5203         113146.500000        12.700000        88809.203125   
5204          43336.500000         0.000000        22876.300781   
...                    ...              ...                 ...   
93019             0.000000         0.000000            0.000000   
93020             0.000000         0.000000            0.000000   
93021             0.000000         0.000000            0.000000   
93022 

In [7]:
display(datasetX.iloc[:,:45].corr().apply(lambda a : a.apply(lambda b: b if b > 0.8 else -10)))

absolute_humidity_2m:gm3_00  \
absolute_humidity_2m:gm3_00                           1.000000   
air_density_2m:kgm3_00                              -10.000000   
ceiling_height_agl:m_00                             -10.000000   
clear_sky_energy_1h:J_00                            -10.000000   
clear_sky_rad:W_00                                  -10.000000   
cloud_base_agl:m_00                                 -10.000000   
dew_or_rime:idx_00                                  -10.000000   
dew_point_2m:K_00                                     0.972589   
diffuse_rad:W_00                                    -10.000000   
diffuse_rad_1h:J_00                                 -10.000000   
direct_rad:W_00                                     -10.000000   
direct_rad_1h:J_00                                  -10.000000   
effective_cloud_cover:p_00                          -10.000000   
elevation:m_00                                      -10.000000   
fresh_snow_12h:cm_00                                -10.000000   
fresh_snow_1h:cm_00                                 -10.000000   
fresh_snow_24h:cm_00                                -10.000000   
fresh_snow_3h:cm_00                                 -10.000000   
fresh_snow_6h:cm_00                                 -10.000000   
is_day:idx_00                                       -10.000000   
is_in_shadow:idx_00                                 -10.000000   
msl_pressure:hPa_00                                 -10.000000   
precip_5min:mm_00                                   -10.000000   
precip_type_5min:idx_00                             -10.000000   
pressure_100m:hPa_00                                -10.000000   
pressure_50m:hPa_00                                 -10.000000   
prob_rime:p_00                                      -10.000000   
rain_water:kgm2_00                                  -10.000000   
relative_humidity_1000hPa:p_00                      -10.000000   
sfc_pressure:hPa_00                                 -10.000000   
snow_density:kgm3_00                                -10.000000   
snow_depth:cm_00                                    -10.000000   
snow_drift:idx_00                                   -10.000000   
snow_melt_10min:mm_00                               -10.000000   
snow_water:kgm2_00                                  -10.000000   
sun_azimuth:d_00                                    -10.000000   
sun_elevation:d_00                                  -10.000000   
super_cooled_liquid_water:kgm2_00                   -10.000000   
t_1000hPa:K_00                                        0.897826   
total_cloud_cover:p_00                              -10.000000   
visibility:m_00                                     -10.000000   
wind_speed_10m:ms_00                                -10.000000   
wind_speed_u_10m:ms_00                              -10.000000   
wind_speed_v_10m:ms_00                              -10.000000   
wind_speed_w_1000hPa:ms_00                          -10.000000   

                                   air_density_2m:kgm3_00  \
absolute_humidity_2m:gm3_00                         -10.0   
air_density_2m:kgm3_00                                1.0   
ceiling_height_agl:m_00                             -10.0   
clear_sky_energy_1h:J_00                            -10.0   
clear_sky_rad:W_00                                  -10.0   
cloud_base_agl:m_00                                 -10.0   
dew_or_rime:idx_00                                  -10.0   
dew_point_2m:K_00                                   -10.0   
diffuse_rad:W_00                                    -10.0   
diffuse_rad_1h:J_00                                 -10.0   
direct_rad:W_00                                     -10.0   
direct_rad_1h:J_00                                  -10.0   
effective_cloud_cover:p_00                          -10.0   
elevation:m_00                                      -10.0   
fresh_snow_12h:cm_00                                -10.0   
fresh_snow_1h:cm_00   

In [14]:
numModels = 5

models = []

for i in range(numModels):
    models.append(tf.keras.models.Sequential([
        #tf.keras.layers.GaussianNoise(stddev=0.1, seed=42),
        tf.keras.layers.Dense(150, activation="tanh"),
        tf.keras.layers.Dense(120, activation="relu"),
        tf.keras.layers.Dense(90, activation="relu"),
        tf.keras.layers.Dense(60, activation="relu"),
        tf.keras.layers.Dense(1, activation="relu"),
    ]))

In [15]:


i = 0
while(i < numModels):
    
    models[i].compile(
        optimizer=tf.keras.optimizers.experimental.Adadelta(learning_rate=1, 
                                                            #weight_decay=0.0001
                                                            ),
        loss="mean_absolute_error"
    )

    history = models[i].fit(x = trainsetX,
                        y = trainsetY,
                        batch_size = 1000,
                        epochs=30,
                        verbose=0
                    )
    
    #retry if model is nonsensical
    loss = models[i].evaluate(evalsetX, evalsetY)
    if(loss > 200):
        continue
    
    print(i, " complete")
    i+=1

313/313 [==============================] - 1s 2ms/step - loss: 139.9128
0  complete
313/313 [==============================] - 1s 3ms/step - loss: 124.5176
1  complete
313/313 [==============================] - 1s 2ms/step - loss: 128.2174
2  complete
313/313 [==============================] - 1s 2ms/step - loss: 137.9295
3  complete
313/313 [==============================] - 3s 6ms/step - loss: 131.3259
4  complete


In [16]:
#!ReWrite
def evaluate_models(models, X, Y):
    preds = X.iloc[:,1:2]

    for i in range(len(models)):
        preds[str(i)] = models[i].predict(X)
        
    
    preds = preds.iloc[:,1:]
    
    
    preds["final"] = preds.mean(axis=1)
    preds["losses"] = (preds["final"] - Y["pv_measurement"]).apply(lambda a : np.abs(a))
    return preds["losses"].mean()


In [17]:

print("entire:")
results = evaluate_models(models ,evalsetX, evalsetY)
print(results)



entire:


313/313 [==============================] - 1s 4ms/step
124.95595654430144


131.8, 30.6, 125.4, 111.9, 41.0, 152.5

124.9, 24.5, 118.7, 107.9, 42.0, 145.3

????  ????

Full features:
125.3, 28.3, 120.6, 106.9, 38.9, 149.8

Dropped features(sfc_pressure:hPa, msl_pressure:hPa, direct_rad_1h:J)
120.3, 28.0, 118.2, 107.9, 38.7, 151.49

Ensamble2:
134.3 29.5 127.1 109.5 40.6

Ensamble2 20 epochs:
128.9 28.5 124.3 108.8 41.3

Ensamble2 20 epochs, fixed nanrows
124.8 29.0 118.9 108.2 40.7

Ensamble2, 20 epochs, dropped location, smaller network, fixed nanrows


Ensamble2 with dropped features(sfc_pressure:hPa, msl_pressure:hPa, direct_rad_1h:J, location) and small network
130.7

Ensamble2 with dropped features(sfc_pressure:hPa, msl_pressure:hPa, direct_rad_1h:J, location) and small network + relu
144

Ensamble2 with 5 models
136.3

Ensamble2 with 5 models 30 epochs
124.9

"Ensamble4":
124.1 28.2 119.5 108.3 41.2 -- 84.26

"Ensamble4" double tanh:
124.1 27.9 120.5 109.5 40.9 -- 84.58

"Ensamble4" 100tanh, 100relu:
123.2 27.8 120.6 107.5 41.7 -- 84.16

"Ensamble4" 100tanh, 100relu, frac=0.5:
122.5 27.8 120.2 108.0 41.9 -- 84.04

Ensamble of Ensambles:
122.5 28.1 119.1 107.7 41,3 -- 83.74

Ensamble of EnsamblesVar1:
122.4


In [ ]:
print("may june july:")
resultsMayJuneJuly = evaluate_models(models ,evalsetX[evalsetIndexMayJuneJuly], evalsetY[evalsetIndexMayJuneJuly])
print(resultsMayJuneJuly)



In [ ]:
print("location A")
resultA = evaluate_models(models ,evalsetX[evalsetX["location_A"].apply(lambda a : a == 1)], 
                         evalsetY[evalsetX["location_A"].apply(lambda a : a == 1)])
print(resultA)

print("location B")
resultB = evaluate_models(models ,evalsetX[evalsetX["location_B"].apply(lambda a : a == 1)], 
                         evalsetY[evalsetX["location_B"].apply(lambda a : a == 1)])
print(resultB)

print("location C")
resultC = evaluate_models(models ,evalsetX[evalsetX["location_C"].apply(lambda a : a == 1)], 
                         evalsetY[evalsetX["location_C"].apply(lambda a : a == 1)])
print(resultC)



In [1]:
print("Small samplesize 1")
resultSmall1 = evaluate_models(models, evalsetX.iloc[:1300,:], evalsetY.iloc[:1300,:])
print(resultSmall1)

print("Small samplesize 2")
resultSmall2 = evaluate_models(models, evalsetX.iloc[1300:2600,:], evalsetY.iloc[1300:2600,:])
print(resultSmall2)

print("Small samplesize 3")
resultSmall3 = evaluate_models(models, evalsetX.iloc[2600:3900,:], evalsetY.iloc[2600:3900,:])
print(resultSmall3)

Small samplesize 1


NameError: name 'evaluate_models' is not defined

67    68

249.9 248.3

90.5  88.2

33.0  36.9

71.5  69.6
152.4  148.7
152.5 148.9
33.2 32.0
17.6 16.9

71.5  72.4  72.4
152.4 153.3 152.7
152.5 155.0 155.0
33.2  33.1  33.1
17.6  17.5  17.5

With 50% training data: tanh
76.5  72.0
161.7 158.6
164.0 164.9
34.6  35.2
18.7  18.7
74.2  74.2
82.1  84.6

With 50% training data: relu


In [86]:
X_test_estimated_a = pd.read_parquet('A/X_test_estimated.parquet')
X_test_estimated_b = pd.read_parquet('B/X_test_estimated.parquet')
X_test_estimated_c = pd.read_parquet('C/X_test_estimated.parquet')

X_test_estimated_a["location"] = "A"
X_test_estimated_b["location"] = "B"
X_test_estimated_c["location"] = "C"

X_test_raw = pd.concat([
                     X_test_estimated_a,
                     X_test_estimated_b,
                     X_test_estimated_c])

#feature indicating time between date_calc and date_forecast
X_test_raw["calc_time"] =(X_test_raw["date_forecast"] - X_test_raw["date_calc"]).astype('timedelta64[s]')

#fill nans
X_test_raw["snow_density:kgm3"] = X_test_raw["snow_density:kgm3"].apply(
    lambda a : np.isnan(a)
    ).map({True: 0, False: 1})
X_test_raw["ceiling_height_agl:m"] = X_test_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)
X_test_raw["cloud_base_agl:m"] = X_test_raw["ceiling_height_agl:m"].apply(
    lambda a : -1000 if np.isnan(a) else a
)

#create seperate dataframes for measurments at minute 00, 15, 30 and 45
X_test00 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 0)].reset_index().iloc[:,1:]
X_test15 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 15)].reset_index().iloc[:,1:]
X_test30 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 30)].reset_index().iloc[:,1:]
X_test45 = X_test_raw[X_test_raw["date_forecast"].apply(lambda time: time.minute == 45)].reset_index().iloc[:,1:]

#remove redundant data
X_test15 = X_test15.iloc[:,2:-2]
X_test30 = X_test30.iloc[:,2:-2]
X_test45 = X_test45.iloc[:,2:-2]

#join observations into single sample
X_test_estimated = X_test00.join(X_test15, lsuffix="_00", rsuffix="_15").join(X_test30.join(X_test45, lsuffix="_30", rsuffix="_45"))


X_test_estimated = X_test_estimated.rename(columns={"date_forecast" : "time"})

parse_dates = ['time']
X_test_targets = pd.read_csv("test.csv", parse_dates=parse_dates)

X_test = pd.merge(X_test_estimated, X_test_targets, on=["time", "location"], how="right").iloc[:,:-2]


X_test["day"] = X_test["time"].dt.day_of_year
X_test["hour"] = X_test["time"].dt.hour

X_test["location_A"] = X_test["location"].apply(lambda a : a == "A").map({True: 1, False: 0})
X_test["location_B"] = X_test["location"].apply(lambda a : a == "B").map({True: 1, False: 0})
X_test["location_C"] = X_test["location"].apply(lambda a : a == "C").map({True: 1, False: 0})

X_test = X_test.drop("location", axis=1)

X_test = X_test.iloc[:,2:]


X_test.iloc[:,:-4] = ((X_test.iloc[:,:-4]-dataMean[:-4])/dataStd[:-4]).fillna(value=0)

display(X_test)



/var/folders/bm/37km6rb530l8y0vv0znc0zb00000gn/T/ipykernel_12419/1064356002.py:63: PerformanceWarning: Adding/subtracting object-dtype array to TimedeltaArray not vectorized.
  X_test.iloc[:,:-4] = ((X_test.iloc[:,:-4]-dataMean[:-4])/dataStd[:-4]).fillna(value=0)


absolute_humidity_2m:gm3_00  air_density_2m:kgm3_00  \
0                       -0.595904                0.834929   
1                       -0.632742                0.862243   
2                       -0.669581                0.780297   
3                       -0.706420                0.725662   
4                       -0.780097                0.725662   
...                           ...                     ...   
2155                     0.840800               -1.623536   
2156                     0.914477               -1.541587   
2157                     1.024993               -1.459637   
2158                     1.098670               -1.350374   
2159                     1.098670               -1.323056   

      ceiling_height_agl:m_00  clear_sky_energy_1h:J_00  clear_sky_rad:W_00  \
0                   -0.382669                 -0.627834           -0.626244   
1                   -1.083697                 -0.627834           -0.626244   
2                   -0.173977                 -0.627834           -0.626244   
3                    0.128249                 -0.627834           -0.626244   
4                    0.134297                 -0.554985           -0.438066   
...                       ...                       ...                 ...   
2155                 0.577483                  0.122098           -0.113350   
2156                 0.539586                 -0.299286           -0.451195   
2157                 0.197264                 -0.540086           -0.604801   
2158                 0.014887                 -0.626155           -0.626244   
2159                 0.017892                 -0.627834           -0.626244   

      cloud_base_agl:m_00  dew_or_rime:idx_00  dew_point_2m:K_00  \
0               -0.382669           -0.028554          -0.517627   
1               -1.083697           -0.028554          -0.532260   
2               -0.173977           -0.028554          -0.590784   
3                0.128249           -0.028554          -0.649314   
4                0.134297           -0.028554          -0.722471   
...                   ...                 ...                ...   
2155             0.577483           -0.028554           0.930882   
2156             0.539586           -0.028554           0.974774   
2157             0.197264           -0.028554           1.018671   
2158             0.014887           -0.028554           1.077196   
2159             0.017892           -0.028554           1.062563   

      diffuse_rad:W_00  diffuse_rad_1h:J_00  direct_rad:W_00  \
0            -0.651235            -0.658524        -0.444504   
1            -0.651235            -0.658524        -0.444504   
2            -0.651235            -0.658524        -0.444504   
3            -0.651235            -0.658524        -0.444504   
4            -0.133490            -0.396493        -0.347113   
...                ...                  ...              ...   
2155          0.029749             0.264405        -0.301073   
2156         -0.326408            -0.149611        -0.392267   
2157         -0.568792            -0.452422        -0.444504   
2158         -0.651235            -0.616985        -0.444504   
2159         -0.651235            -0.658524        -0.444504   

      direct_rad_1h:J_00  effective_cloud_cover:p_00  elevation:m_00  \
0              -0.449899                    0.390542       -0.685740   
1              -0.449899                   -0.071723       -0.685740   
2              -0.449899                    0.781470       -0.685740   
3              -0.449899                    0.227893       -0.685740   
4              -0.400659                   -0.240079       -0.685740   
...                  ...                         ...             ...   
2155           -0.307607                    0.644503        1.599325   
2156           -0.350643                    0.496121        1.599325   
2157           -0.423319                    0.050977        1.599325   
2158           -0.

In [87]:
#!ReWrite
def models_predict(models, X):
    preds = X.iloc[:,1:2]

    for i in range(len(models)):
        preds[str(i)] = models[i].predict(X)

    preds = preds.iloc[:,1:]
    preds["final"] = preds.mean(axis=1)
    return preds["final"]

In [89]:
test_preds = pd.read_csv("sample_submission.csv")
test_preds["prediction"] = models_predict(models, X_test)
display(test_preds)
test_preds.to_csv("preds/Ensamble3.csv", index=False)

68/68 [==============================] - 0s 2ms/step


id  prediction
0        0    0.000000
1        1    0.000000
2        2    0.000000
3        3   53.867607
4        4  388.250275
...    ...         ...
2155  2155   53.368217
2156  2156   22.887728
2157  2157    0.000000
2158  2158    0.000000
2159  2159    0.000000

[2160 rows x 2 columns]